pallets/router/src/mock.rs

```diff
@@ -0,0 +1,234 @@
+}
+
+impl pallet_amm::Config for Runtime {
+    type Event = Event;
+    type Currency = Currencies;
+    type PalletId = AMMPalletId;
+    type WeightInfo = ();
+    type AllowPermissionlessPoolCreation = AllowPermissionlessPoolCreation;
+    type LpFee = DefaultLpFee;
+    type ProtocolFee = DefaultProtocolFee;
+    type ProtocolFeeReceiver = DefaultProtocolFeeReceiver;
+}
+
+parameter_types! {
+    pub const MaxLengthRoute: u8 = 10;
+    pub Routes: Route = vec![
+        (0, DOT, XDOT),
+        (1, XDOT, DOT),
+        (2, XDOT, CurrencyId::Token(TokenSymbol::KSM)),
+        (3, CurrencyId::Token(TokenSymbol::KSM), XDOT),
+        (4, CurrencyId::Token(TokenSymbol::xDOT), XDOT),
+        (5, XDOT, CurrencyId::Token(TokenSymbol::xDOT)),
+    ];
+}
+
+impl Config for Runtime {
+    type Event = Event;
+    type RoutePalletId = RoutePalletId;
+    type AMMAdaptor = DOT2XDOT;
+    type Routes = Routes;
+    type MaxLengthRoute = MaxLengthRoute;
+    type Currency = Currencies;
+}
+
+type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Runtime>;
+type Block = frame_system::mocking::MockBlock<Runtime>;
+
+construct_runtime!(
+    pub enum Runtime where
+        Block = Block,
+        NodeBlock = Block,
+        UncheckedExtrinsic = UncheckedExtrinsic
+    {
+        System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
+        Balances: pallet_balances::{Pallet, Call, Storage, Event<T>},
+        Tokens: orml_tokens::{Pallet, Storage, Config<T>, Event<T>},
+        Currencies: orml_currencies::{Pallet, Call, Event<T>},
+        // AMM instances
+        DOT2XDOT: pallet_amm::{Pallet, Call, Storage, Event<T>},
+        // AMM Route
+        AMMRoute: pallet_route::{Pallet, Call, Event<T>},
+    }
+);
+
+pub struct ExtBuilder {
+    pub balances: Vec<(AccountId, CurrencyId, Balance)>,
+}
+
+impl Default for ExtBuilder {
+    fn default() -> Self {
+        Self {
+            balances: vec![
+                (ALICE.into(), DOT, 10_000),
+                (BOB.into(), DOT, 10_000),
+                (DAVE.into(), DOT, 1_000_000_000),
+                (DAVE.into(), XDOT, 1_000_000_000),
+                (BOB.into(), XDOT, 5_000),
+            ],
+        }
+    }
+}
+
+impl ExtBuilder {
+    pub fn build(self) -> sp_io::TestExternalities {
+        let mut t = frame_system::GenesisConfig::default()
+            .build_storage::<Runtime>()
+            .unwrap();
+
+        orml_tokens::GenesisConfig::<Runtime> {
+            balances: self.balances.clone(),
+        }
+        .assimilate_storage(&mut t)
+        .unwrap();
+        t.into()
+    }
+}
+
+pub(crate) fn run_to_block(n: u64) {
+    while System::block_number() < n {
+        AMMRoute::on_finalize(System::block_number());
+        System::on_finalize(System::block_number());
+        System::set_block_number(System::block_number() + 1);
+        System::on_initialize(System::block_number());
+        AMMRoute::on_initialize(System::block_number());
+    }
+}
```

Comment:
we don't rely on the `on_initialize` and `on_finalize` functions for this pallet. Thus we should be able to replace this as below:
```suggestion
pub(crate) fn run_to_block(n: u64) {
    System::set_block_number(n);
}
```